pkg/openapiconv/convert.go

```diff
@@ -0,0 +1,270 @@
+	components := &spec3.Components{
+		Schemas:         make(map[string]*spec.Schema),
+		SecuritySchemes: make(spec3.SecuritySchemes),
+		Responses:       make(map[string]*spec3.Response),
+	}
+	for s, schema := range v2Definitions {
+		components.Schemas[s] = ConvertSchema(&schema)
+	}
+
+	for s, securityScheme := range v2SecurityDefinitions {
+		components.SecuritySchemes[s] = ConvertSecurityScheme(securityScheme)
+	}
+	for r, response := range v2Responses {
+		components.Responses[r] = ConvertResponse(&response, produces)
+	}
+
+	return components
+}
+
+func ConvertSchema(v2Schema *spec.Schema) *spec.Schema {
+	if v2Schema == nil {
+		return nil
+	}
+	var v3Schema spec.Schema
+	v3Schema = *v2Schema
+
+	if refString := v2Schema.Ref.String(); refString != "" {
+		if idx := strings.Index(refString, OpenAPIV2DefPrefix); idx != -1 {
+			v3Schema.Ref = spec.MustCreateRef(OpenAPIV3DefPrefix + refString[idx+len(OpenAPIV2DefPrefix):])
+		} else {
+			klog.Errorf("Error: Swagger V2 Ref %s does not contain #/definitions\n", refString)
+		}
+	}
+
+	for key, property := range v2Schema.Properties {
+		v3Schema.Properties[key] = *ConvertSchema(&property)
+	}
+	if v2Schema.Items != nil {
+		v3Schema.Items.Schema = ConvertSchema(v2Schema.Items.Schema)
+		if len(v2Schema.Items.Schemas) > 0 {
+			v3Schema.Items.Schemas = []spec.Schema{}
+			for _, s := range v2Schema.Items.Schemas {
+				v3Schema.Items.Schemas = append(v3Schema.Items.Schemas, *ConvertSchema(&s))
+			}
+		}
+	}
+
+	if v2Schema.AdditionalProperties != nil {
+		v3Schema.AdditionalProperties.Schema = ConvertSchema(v2Schema.AdditionalProperties.Schema)
+	}
+	if v2Schema.AdditionalItems != nil {
+		v3Schema.AdditionalProperties.Schema = ConvertSchema(v2Schema.AdditionalProperties.Schema)
+	}
+
+	return &v3Schema
+}
+
+func ConvertSecurityScheme(v2securityScheme *spec.SecurityScheme) *spec3.SecurityScheme {
+	if v2securityScheme == nil {
+		return nil
+	}
+	securityScheme := &spec3.SecurityScheme{
+		VendorExtensible: v2securityScheme.VendorExtensible,
+		SecuritySchemeProps: spec3.SecuritySchemeProps{
+			Description: v2securityScheme.Description,
+			Type:        v2securityScheme.Type,
+			Name:        v2securityScheme.Name,
+			In:          v2securityScheme.In,
+		},
+	}
+
+	if v2securityScheme.Flow != "" {
+		securityScheme.Flows = make(map[string]*spec3.OAuthFlow)
+		securityScheme.Flows[v2securityScheme.Flow] = &spec3.OAuthFlow{
+			OAuthFlowProps: spec3.OAuthFlowProps{
+				AuthorizationUrl: v2securityScheme.AuthorizationURL,
+				TokenUrl:         v2securityScheme.TokenURL,
+				Scopes:           v2securityScheme.Scopes,
+			},
+		}
+	}
+	return securityScheme
+}
+
+func ConvertPaths(v2Paths *spec.Paths) *spec3.Paths {
+	if v2Paths == nil {
+		return nil
+	}
+	paths := &spec3.Paths{
+		VendorExtensible: v2Paths.VendorExtensible,
+		Paths:            make(map[string]*spec3.Path),
```

Comment:
`Paths` should be non-nil if and only if `v2Paths.Paths` is non-nil. This same reasoning also applies elsewhere: if `spec` lists the fields as `nil`, `spec3` should also use `nil`. If `spec` is an empty map/list, `spec3` should be an empty map/list.

Reply:
Hm, is there a significance between nil vs empty maps? They would json marshal to the same result but I guess their representation is different in golang. I'm sure there are a bunch of other instances of this in the code, is there a reason we'd prefer one over the other?

Reply:
Are we 100% sure that they serialize to the same thing? There's clearly a difference between a missing map and an empty map (s/map/list/)

Reply:
They serialize to the same thing but reflect.DeepEqual will fail the comparison. I've added length checks to avoid initializing 0 length maps/lists.